File: app/src/main/java/com/nruge/iceinfo/MenuRepository.kt
package com.nruge.iceinfo

import android.util.Log
import com.nruge.iceinfo.model.AvailabilityItem
import com.nruge.iceinfo.model.MenuCategory
import com.nruge.iceinfo.model.MenuDeclarationEntry
import com.nruge.iceinfo.model.MenuPageResponse
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

data class MenuResult(
    val categories: List<MenuCategory>,
    val declarations: Map<String, String>
)

object MenuRepository {

    private const val API_PATH = "/bap/api/products"
    private const val AVAILABILITY_PATH = "/bap/api/availabilities"
    private val hosts = listOf("https://iceportal.de", "http://iceportal.de")

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) {
            json(json, contentType = io.ktor.http.ContentType.Any)
        }
        install(Logging) { level = LogLevel.INFO }
        install(HttpTimeout) {
            requestTimeoutMillis = 8000
            connectTimeoutMillis = 5000
            socketTimeoutMillis = 8000
        }
        defaultRequest { header("Accept", "application/json") }
    }

    suspend fun fetchMenu(): MenuResult = withContext(Dispatchers.IO) {
        for (host in hosts) {
            try {
                val page = client.get("$host$API_PATH").body<MenuPageResponse>()
                val order = com.nruge.iceinfo.model.MENU_CATEGORY_ORDER
                val categories = page.teaserGroups
                    .filter { it.items.isNotEmpty() }
                    .sortedBy { g -> order.indexOf(g.title).let { if (it < 0) Int.MAX_VALUE else it } }
                    .map { MenuCategory(it.title, it.items) }
                val declarations = page.declarationGroup?.items
                    ?.associate { it.key to it.text } ?: emptyMap()
                return@withContext MenuResult(categories, declarations)
            } catch (e: Exception) {
                Log.w("MenuRepo", "GET $host$API_PATH failed: ${e.message}")
            }
        }
        MenuResult(emptyList(), emptyMap())
    }

    suspend fun fetchAvailabilities(): Map<Int, Boolean> = withContext(Dispatchers.IO) {
        for (host in hosts) {
            try {
                val items = client.get("$host$AVAILABILITY_PATH").body<List<AvailabilityItem>>()
                return@withContext items.associate { it.ecmId to it.visible }
            } catch (e: Exception) {
                Log.w("MenuRepo", "GET $host$AVAILABILITY_PATH failed: ${e.message}")
            }
        }
        emptyMap()
    }
}

File: ios/ICEInfo/Network/MenuRepository.swift
import Foundation

actor MenuRepository {
    static let shared = MenuRepository()

    private let apiPath = "/bap/api/products"
    private let availabilityPath = "/bap/api/availabilities"
    private let hosts = ["https://iceportal.de", "http://iceportal.de"]

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    func fetchMenu() async -> MenuResult {
        for host in hosts {
            guard let url = URL(string: "\(host)\(apiPath)") else { continue }
            do {
                let request = URLRequest(url: url, timeoutInterval: 8)
                let (data, _) = try await URLSession.shared.data(for: request)
                let page = try decoder.decode(MenuPageResponse.self, from: data)
                let categories = page.teaserGroups
                    .filter { !$0.items.isEmpty }
                    .sorted { a, b in
                        let orderA = menuCategoryOrder.firstIndex(of: a.title) ?? Int.max
                        let orderB = menuCategoryOrder.firstIndex(of: b.title) ?? Int.max
                        return orderA < orderB
                    }
                    .map { MenuCategory(title: $0.title, items: $0.items) }
                let declarations = page.declarationGroup?.items
                    .reduce(into: [:]) { $0[$1.key] = $1.text } ?? [:]
                return MenuResult(categories: categories, declarations: declarations)
            } catch {
                continue
            }
        }
        return MenuResult(categories: [], declarations: [:])
    }

    func fetchAvailabilities() async -> [Int: Bool] {
        for host in hosts {
            guard let url = URL(string: "\(host)\(availabilityPath)") else { continue }
            do {
                let request = URLRequest(url: url, timeoutInterval: 8)
                let (data, _) = try await URLSession.shared.data(for: request)
                let items = try decoder.decode([AvailabilityItem].self, from: data)
                return items.reduce(into: [:]) { $0[$1.ecmId] = $1.visible }
            } catch {
                continue
            }
        }
        return [:]
    }

    nonisolated func applyAvailabilities(categories: [MenuCategory], availabilities: [Int: Bool]) -> [MenuCategory] {
        guard !availabilities.isEmpty else { return categories }
        return categories.map { cat in
            MenuCategory(title: cat.title, items: cat.items.map { item in
                var copy = item
                if let visible = availabilities[item.id] { copy.visible = visible }
                return copy
            })
        }
    }
}
